blas/blas32/conv_test.go

```diff
@@ -0,0 +1,620 @@
+
+type general interface {
+	dims() (r, c int)
+	at(i, j int) float32
+}
+
+func equalGeneral(a, b general) bool {
+	ar, ac := a.dims()
+	br, bc := b.dims()
+	if ar != br || ac != bc {
+		return false
+	}
+	for i := 0; i < ar; i++ {
+		for j := 0; j < ac; j++ {
+			if a.at(i, j) != b.at(i, j) || math.IsNaN(a.at(i, j)) != math.IsNaN(b.at(i, j)) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+var generalTests = []General{
+	{Rows: 2, Cols: 3, Stride: 3, Data: []float32{
+		1, 2, 3,
+		4, 5, 6,
+	}},
+	{Rows: 3, Cols: 2, Stride: 2, Data: []float32{
+		1, 2,
+		3, 4,
+		5, 6,
+	}},
+	{Rows: 3, Cols: 3, Stride: 3, Data: []float32{
+		1, 2, 3,
+		4, 5, 6,
+		7, 8, 9,
+	}},
+	{Rows: 2, Cols: 3, Stride: 5, Data: []float32{
+		1, 2, 3, 0, 0,
+		4, 5, 6, 0, 0,
+	}},
+	{Rows: 3, Cols: 2, Stride: 5, Data: []float32{
+		1, 2, 0, 0, 0,
+		3, 4, 0, 0, 0,
+		5, 6, 0, 0, 0,
+	}},
+	{Rows: 3, Cols: 3, Stride: 5, Data: []float32{
+		1, 2, 3, 0, 0,
+		4, 5, 6, 0, 0,
+		7, 8, 9, 0, 0,
+	}},
+}
+
+func TestConvertGeneral(t *testing.T) {
+	for _, test := range generalTests {
+		colmajor := newGeneralColsFrom(test)
+		if !equalGeneral(colmajor, test) {
+			t.Errorf("unexpected result for row major to col major conversion:\n\tgot: %#v\n\tfrom:%#v",
+				colmajor, test)
+		}
+		rowmajor := newGeneralFrom(colmajor)
+		if !equalGeneral(rowmajor, test) {
+			t.Errorf("unexpected result for col major to row major conversion:\n\tgot: %#v\n\twant:%#v",
+				rowmajor, test)
+		}
+	}
+}
+
+func newTriangularFrom(a TriangularCols) Triangular {
+	t := Triangular{
+		N:      a.N,
+		Stride: a.N,
+		Data:   make([]float32, a.N*a.N),
+		Diag:   a.Diag,
+		Uplo:   a.Uplo,
+	}
+	t.From(a)
+	return t
+}
+
+func (m Triangular) n() int { return m.N }
+func (m Triangular) at(i, j int) float32 {
+	if m.Diag == blas.Unit && i == j {
+		return 1
+	}
+	if m.Uplo == blas.Lower && i < j && j < m.N {
+		return 0
+	}
+	if m.Uplo == blas.Upper && i > j {
+		return 0
+	}
+	return m.Data[i*m.Stride+j]
+}
+func (m Triangular) uplo() blas.Uplo { return m.Uplo }
+func (m Triangular) diag() blas.Diag { return m.Diag }
+
+func newTriangularColsFrom(a Triangular) TriangularCols {
+	t := TriangularCols{
+		N:      a.N,
+		Stride: a.N,
+		Data:   make([]float32, a.N*a.N),
+		Diag:   a.Diag,
+		Uplo:   a.Uplo,
+	}
+	t.From(a)
+	return t
+}
+
+func (m TriangularCols) n() int { return m.N }
+func (m TriangularCols) at(i, j int) float32 {
+	if m.Diag == blas.Unit && i == j {
+		return 1
+	}
+	if m.Uplo == blas.Lower && i < j {
+		return 0
+	}
+	if m.Uplo == blas.Upper && i > j && i < m.N {
+		return 0
+	}
+	return m.Data[i+j*m.Stride]
+}
+func (m TriangularCols) uplo() blas.Uplo { return m.Uplo }
+func (m TriangularCols) diag() blas.Diag { return m.Diag }
+
+type triangular interface {
+	n() int
+	at(i, j int) float32
+	uplo() blas.Uplo
+	diag() blas.Diag
+}
+
+func equalTriangular(a, b triangular) bool {
+	an := a.n()
+	bn := b.n()
+	if an != bn {
+		return false
+	}
+	for i := 0; i < an; i++ {
```

Comment:
It doesn't really matter since we are only dealing with matrix elements, not matrix metadata.